Sometimes panics on startup

Hello,

I noticed the following panic on startup:

```
{"level":"info","time":1648997805,"message":"creating kubernetes clientset"}
panic: runtime error: index out of range [0] with length 0

goroutine 9 [running]:
main.(*Collector).Collect(0xc00000f320, 0x0)
	/opt/caexporter/caexporter.go:95 +0x6a5
github.com/prometheus/client_golang/prometheus.DescribeByCollect.func1()
	/go/pkg/mod/github.com/prometheus/client_golang@v1.11.0/prometheus/collector.go:90 +0x2b
created by github.com/prometheus/client_golang/prometheus.DescribeByCollect
	/go/pkg/mod/github.com/prometheus/client_golang@v1.11.0/prometheus/collector.go:89 +0xa5
```

I think it's because `prometheus` tries to scrape from the exporter but it hasn't yet finished initialization. I suggest implementing a `ready` endpoint to avoid premature scraps

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,6 +1,6 @@
-FROM golang:1.17-alpine as builder
+FROM golang:1.19-alpine as builder
 
-ARG caexporter_version=0.0.1
+ARG caexporter_version=0.0.2
 
 RUN apk --no-cache add git
 
@@ -8,7 +8,7 @@ WORKDIR /opt/caexporter
 COPY . .
 RUN go build -ldflags="-X github.com/prometheus/common/version.Version=${caexporter_version}} -X github.com/prometheus/common/version.Revision=$(git rev-parse HEAD) -X github.com/prometheus/common/version.Branch=$(git branch --show-current)" ./
 
-FROM alpine:3.15
+FROM alpine:3.16
 LABEL maintainer "Bringg DevOps <devops@bringg.com>"
 
 COPY --from=builder /opt/caexporter/caexporter /usr/local/bin

--- a/caexporter.go
+++ b/caexporter.go
@@ -36,9 +36,9 @@ var (
 	)
 
 	scrapeError = prometheus.NewGauge(prometheus.GaugeOpts{
-		Namespace: namespace, 
-		Name: "scrape_error",
-		Help: "Set to 0 for successful scrape, or 1 otherwise",	
+		Namespace: namespace,
+		Name:      "scrape_error",
+		Help:      "Set to 0 for successful scrape, or 1 otherwise",
 	})
 )
 
@@ -61,14 +61,17 @@ func newCollector() *Collector {
 		log.Fatal().Err(err).Msg("")
 	}
 
-	log.Info().Msg("creating kubernetes clientset")
+	log.Info().Msg("Creating kubernetes clientset")
 
 	clientset, err := kubernetes.NewForConfig(cfg)
 	if err != nil {
 		log.Fatal().Err(err).Msg("")
 	}
 
-	rx := regexp.MustCompile(`(?s)LastProbeTime:\s*(.{39})`)
+	// we're looking for lines that look like -
+	// LastProbeTime:      2022-09-11 11:21:27.046154458 +0000 UTC m=+25121.073170634
+	// but without the m=... part
+	rx := regexp.MustCompile(`(?s)LastProbeTime:\s*([\d\-]*\s[\d\:\.]*\s[\+\d]*\sUTC)`)
 
 	c := &Collector{
 		kubeClient: clientset,
@@ -97,6 +100,14 @@ func (c *Collector) Collect(ch chan<- prometheus.Metric) {
 	cmData := cm.Data["status"]
 
 	cw := c.regex.FindAllStringSubmatch(cmData, 3)
+	// nil means there's no regex match on the returned data,
+	// for example the configmap is still in initialized phase
+	if cw == nil {
+		log.Info().Msg("Can't match regex pattern on the returned data")
+		log.Debug().Msgf("Received configmap status %s, regex pattern is %s", cmData, c.regex.String())
+		scrapeError.Set(1)
+		return
+	}
 
 	res := map[string]string{
 		"main":      cw[0][1],
@@ -146,11 +157,13 @@ func main() {
 	// log configs
 	zerolog.TimeFieldFormat = zerolog.TimeFormatUnix
 
+	zerolog.SetGlobalLevel(zerolog.InfoLevel)
 	if *logDebug {
 		zerolog.SetGlobalLevel(zerolog.DebugLevel)
-	} else {
-		zerolog.SetGlobalLevel(zerolog.InfoLevel)
 	}
+
+	// adds line location to the caller field of the event
+	log.Logger = log.With().Caller().Logger()
 
 	prometheus.MustRegister(newCollector())
 

--- a/go.mod
+++ b/go.mod
@@ -1,6 +1,6 @@
 module github.com/bringg/caexporter
 
-go 1.17
+go 1.19
 
 require (
 	github.com/prometheus/client_golang v1.11.0

